onnxruntime/python/tools/transformers/parity_check_helper.py

```diff
@@ -0,0 +1,113 @@
+    os.environ["ORT_DEBUG_NODE_IO_OUTPUT_DIR"] = output_path
+
+def environ_reset():
+    for flag in ["ORT_DEBUG_NODE_IO_DUMP_SHAPE_DATA",
+                 "ORT_DEBUG_NODE_IO_DUMP_INPUT_DATA",
+                 "ORT_DEBUG_NODE_IO_DUMP_OUTPUT_DATA",
+                 "ORT_DEBUG_NODE_IO_NAME_FILTER",
+                 "ORT_DEBUG_NODE_IO_OP_TYPE_FILTER",
+                 "ORT_DEBUG_NODE_IO_DUMP_DATA_TO_FILES",
+                 "ORT_DEBUG_NODE_IO_OUTPUT_DIR",
+                 "ORT_DEBUG_NODE_IO_DUMPING_DATA_TO_FILES_FOR_ALL_NODES_IS_OK"]:
+        if flag in os.environ:
+            del os.environ[flag]
+
+def inference(model_path, dummy_inputs, outputs_path, use_gpu):
+    environ_reset()
+    environ_setting_nodes()
+    environ_setting_paths(outputs_path)
+    session = create_onnxruntime_session(model_path, use_gpu, enable_all_optimization=False)
+    Gpt2Helper.onnxruntime_inference(session, dummy_inputs)
+
+def generate_outputs_files(model_path, dummy_inputs, outputs_path, use_gpu):
+    dir_path = Path(outputs_path)
+    if dir_path.exists() and dir_path.is_dir():
+        import shutil
+        shutil.rmtree(outputs_path)
+    dir_path.mkdir(parents=True, exist_ok=True)
+
+    process = multiprocessing.Process(target=inference,
+                                      args=(model_path, dummy_inputs, outputs_path, use_gpu))
+    process.start()
+    process.join()
+
+def post_processing(outputs_path, outputs_path_other):
+    # Compare outputs with e.g. fp16 and fp32
+    record = {}
+    if_close = {}
+
+    import glob
+    for filename in glob.glob(os.path.join(outputs_path, '*.tensorproto')):
+      filename_other = os.path.join(outputs_path_other, Path(filename).name)
+      if not os.path.exists(filename_other):
+          continue
+      with open(filename, 'rb') as f:
+        tensor = TensorProto()
+        tensor.ParseFromString(f.read())
+        array = numpy_helper.to_array(tensor)
+        with open(filename_other, 'rb') as f:
+          tensor_other = TensorProto()
+          tensor_other.ParseFromString(f.read())
+          array_other = numpy_helper.to_array(tensor_other)
+          if array_other.size == 0:
+            continue
+          diff = numpy.average(numpy.abs(array_other - array)/(numpy.abs(array_other) + 1e-6))
+          if math.isnan(diff):
+            continue
+          record[Path(filename).name.split(".")[0]] = diff
+          if_close[Path(filename).name.split(".")[0]] = numpy.allclose(array, array_other, rtol=1e-04, atol=1e-04)
+
+    results = [f"Node\tDiff\tClose"]
+    for k, v in sorted(record.items(), key=lambda x: x[1], reverse=True):
+        results.append(f"{k}\t{v}\t{if_close[k]}")
+    for line in results:
+        print(line)
+
+if __name__ == '__main__':
+    multiprocessing.set_start_method('spawn')
+    print("Prepare inputs, models and compare")
+    # input_fp16 = Gpt2Helper.get_dummy_inputs(...), input_fp32 = inputs_fp16.to_fp32()
```

Comment:
Could you implement this part for comparing fp16 and fp32 GPT2 model? That could serve as an example.

Reply:
Sure